charts/spire/charts/spire-server/values.yaml

```diff
@@ -327,6 +327,37 @@ upstreamAuthority:
       address: ""
       ## @param upstreamAuthority.spire.server.port Port for upstream Spire server
       port: 8081
+  vault:
+    ## @param upstreamAuthority.vault.enabled Enable Hashicorp Vault as upstream CA
+    enabled: false
+    ## @param upstreamAuthority.vault.vaultAddr The URL of the Vault server. (e.g., https://vault.example.com:8443/)
+    vaultAddr: ""
+    ## @param upstreamAuthority.vault.namespace Name of the Vault namespace. This is only available in the Vault Enterprise.
+    namespace: ""
+    ## @param upstreamAuthority.vault.pkiMountPoint Name of the mount point where PKI secret engine is mounted
+    pkiMountPoint: "pki"
+      ## @param upstreamAuthority.vault.insecureSkipVerify If true, caCert options are ignored and Spire accepts any server certificates claiming to be Vault
+    insecureSkipVerify: false
+    caCert:
+      ## @param upstreamAuthority.vault.caCert.type Type of resource representing the Vault server certificate, options are 'Secret' or 'Configmap', the item must be named `ca.crt`
+      type: Secret
+      ## @param upstreamAuthority.vault.caCert.name Name of the Kubernetes resource containing the Vault server certificate
+      name: vault-ca
+
+    k8sAuth:
+      ## @param upstreamAuthority.vault.k8sAuth.enabled Enable k8s authentication to Hashicorp Vault
+      enabled: false
+      ## @param upstreamAuthority.vault.k8sAuth.k8sAuthMountPoint Name of the mount point where the Kubernetes auth method is mounted
+      k8sAuthMountPoint: "kubernetes"
+      ## @param upstreamAuthority.vault.k8sAuth.k8sAuthRoleName Required - Name of the Vault role. The plugin authenticates against the named role
+      k8sAuthRoleName: ""
+      token:
+        ## @param upstreamAuthority.vault.k8sAuth.token.audience Intended audience of the PSAT, it must match one of the audiences supported by the Kubernetes API server. If no audience is specified, it defaults to the identifier of API Server.
```

Comment:
Should tokens intended only to be handed to the kubernetes api server be handed to vault? PSAT allows you to specify the audience to be specific to vault so that it can't be used to talk to kubernetes too. I think we probably should be defaulting the audience to vault and not recommending they match audiences with the api-server?

Reply:
As I understand it, the Vault k8s auth method takes a SA JWT, then validates this SA JWT via the k8s TokenReview API, before associating Vault policies & roles for the requested resource.

If we have a custom audience i.e. `audiences: [vault]`, we must also have that same audience configured on the k8s API-Server via provided flags. If not, when Vault calls the `TokenReview` API, it will fail to validate the JWT as it wont be equal to one of the string elements in the `audiences` array.

For this reason, I decided to have template explicitly use the default value, as I am unsure of what audiences a helm operator would initially have. Further, in a scenario where the helm operator does not necessarily manage their k8s cluster, the default may be a useful option to enable easy configuration with Vault.

Reference:
https://developer.hashicorp.com/vault/docs/auth/kubernetes

Reply:
Thanks for the documentation. It helped a bit...

I see the projection here:
https://github.com/spiffe/helm-charts/pull/415/files#diff-d24749b40abb8bfd0633e64bc366aa9803df1473594e446177244f6bcfb316deR385

but the audience is optional. Which is odd to me for a projected token. I understand reading the docs that vault is doing extra k8s validation so needs it to be in the apiserver audience list of the api server. Is it ever the right thing though to use the apiserver token directly rather then a projected token for the vault service? Feels like you wouldn't be able to easily revoke the vault client should you need to.

So, it still makes sense to me to always project to some client value. If it is the wrong value for the user, they can always configure it themselves? So, an audience of something like 'vault' as a default feels like maybe the right thing to do?

In the case where they don't control the apiserver, they could still set the audience back to 'apiserver' manually to support that case. But more explicitly accepting the tradeoffs that need to be made in that case? Yeah, its a little harder to use, but also more explicit that they are taking the risks?

Reply:
Agreed, I will default the audience to `vault` now